cyral/resource_cyral_integration_idp_saml.go

```diff
@@ -0,0 +1,165 @@
+package cyral
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+
+	"github.com/cyralinc/terraform-provider-cyral/client"
+)
+
+type CreateGenericSAMLRequest struct {
+	SAMLDraftId string                  `json:"samlDraftId"`
+	IdpMetadata *GenericSAMLIdpMetadata `json:"idpMetadata,omitempty"`
+
+	// Currently unused fields
+	IdpDescriptor *GenericSAMLIdpDescriptor `json:"idpDescriptor,omitempty"`
+}
+
+func (req *CreateGenericSAMLRequest) ReadFromSchema(d *schema.ResourceData) error {
+	req.SAMLDraftId = d.Get("saml_draft_id").(string)
+	if url := d.Get("idp_metadata_url").(string); url != "" {
+		req.IdpMetadata = &GenericSAMLIdpMetadata{
+			URL: url,
+		}
+	} else if xml := d.Get("idp_metadata_document").(string); xml != "" {
+		req.IdpMetadata = &GenericSAMLIdpMetadata{
+			XML: xml,
+		}
+	}
+	return nil
+}
+
+type CreateGenericSAMLResponse struct {
+	Integration GenericSAMLIntegration `json:"integration"`
+}
+
+func (resp *CreateGenericSAMLResponse) WriteToSchema(d *schema.ResourceData) error {
+	return resp.Integration.WriteToSchema(d)
+}
+
+func (resp *CreateGenericSAMLResponse) ReadFromSchema(d *schema.ResourceData) error {
+	return resp.Integration.ReadFromSchema(d)
+}
+
+type ReadGenericSAMLResponse struct {
+	IdentityProvider GenericSAMLIntegration `json:"identityProvider"`
+}
+
+func (resp *ReadGenericSAMLResponse) WriteToSchema(d *schema.ResourceData) error {
+	return resp.IdentityProvider.WriteToSchema(d)
+}
+
+func (resp *ReadGenericSAMLResponse) ReadFromSchema(d *schema.ResourceData) error {
+	return resp.IdentityProvider.ReadFromSchema(d)
+}
+
+type UpdateGenericSAMLRequest struct {
+	ID               string                 `json:"id"`
+	IdentityProvider GenericSAMLIntegration `json:"identityProvider"`
+}
+
+func (resp *UpdateGenericSAMLRequest) ReadFromSchema(d *schema.ResourceData) error {
+	resp.ID = d.Id()
+	return resp.IdentityProvider.ReadFromSchema(d)
+}
```

Comment:
The ID is the same of the resource ID and I assume that it cannot be updated once the resource is already created, so we probably can just remove the ID and use a single structure for both Update and Read operations (See my comment above).
```suggestion
```

Reply:
I believe having this separation makes the code easier to read, and it resembles more the API, which might facilitate adaptation of the provider in case of API changes in the future (see reply to similar comment).

Reply:
I see, I think that makes sense. I would just remove the `WriteToSchema` and `ReadFromSchema` methods that are not actually being used.